Orleans.Reminders.DynamoDB use the Query operation for ReadRows

smiron opened the pull request:
Fixes: https://github.com/dotnet/orleans/issues/7657

The `scan` operation is inefficient especially as the table size increases. This is because the `scan` operation reads the entire table (therefore consumes RCUs for all the data in the table) and filters out rows based on the expression.

On the other hand, the `query` operation is very efficient as it pulls out only the specific records that match the HashKey and the SortKey without wasting any RCUs.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/7658)

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=7658) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=7658) before we can accept your contribution.<br/><br/>:x: smiron [sign now](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=7658)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/orleans?pullRequest=7658) it.</sub>

highlyunavailable commented:
How does this handle existing tables that don't have the reminder hash key?

ReubenBond commented on src/AWS/Shared/Storage/DynamoDBStorage.cs:
```diff
@@ -163,34 +165,140 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+            {
+                ret = await GetTableDescription(tableName);
+
+                await Task.Delay(delay);
```
This performs an unconditional wait, even on success. Should it check the loop condition before waiting?

ReubenBond commented:
cc @galvesribeiro, please take a look.
Everyone else is invited to take a look, too.
I have not tested this manually. Is someone able to?

jsteinich commented on src/AWS/Shared/Storage/DynamoDBStorage.cs:
```diff
@@ -163,34 +165,143 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+            }
+        }
+
+        private async Task TableUpdateTtlAsync(string tableName, string ttlAttributeName)
```
Nit: This could return `ValueTask` since it's pretty common that a ttl attribute isn't set.

smiron replied:
Interesting. I'll look into it.

galvesribeiro replied:
I Agree.

jsteinich commented on src/AWS/Shared/Storage/DynamoDBStorage.cs:
```diff
@@ -163,34 +165,143 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+
+        private async Task UpdateTableAsync(TableDescription tableDescription, List<AttributeDefinition> attributes, List<GlobalSecondaryIndex> secondaryIndexes = null, string ttlAttributeName = null)
+        {
+            var request = new UpdateTableRequest
```
Nit: The request object doesn't need to be created if nothing changes. Does make the check for changes a bit more complicated though.

smiron replied:
That is exactly why I left it in place. I'm compiling the desired state using common logic and then just comparing the desired state with the actual state.

I'd prefer to leave this as is if you don't have a strong opinion.

jsteinich commented on src/AWS/Shared/Storage/DynamoDBStorage.cs:
```diff
@@ -163,34 +165,143 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+                return;
+            }
+
+            await ddbClient.UpdateTimeToLiveAsync(new UpdateTimeToLiveRequest
```
Could check the existing ttl attribute on the table. Since it requires a request, hard to say if it's any better though.

jsteinich replied:
I think it's fine to keep as is.

jsteinich commented on src/AWS/Shared/Storage/DynamoDBStorage.cs:
```diff
@@ -163,34 +165,143 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+                        {
+                            IndexName = secondaryIndex.IndexName,
+                            Projection = secondaryIndex.Projection,
+                            ProvisionedThroughput = secondaryIndex.ProvisionedThroughput,
```
ProvisionedThroughput should be taken from the parent table configuration.

smiron replied:
Good spot!

jsteinich commented on src/AWS/Shared/Storage/DynamoDBStorage.cs:
```diff
@@ -163,34 +165,143 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+                AttributeDefinitions = attributes
+            });
+
+            await TableIndexWaitOnStatusAsync(tableName, secondaryIndex.IndexName, IndexStatus.CREATING, IndexStatus.ACTIVE);
```
I'm not certain if it's necessary, but in some similar code, we also check that the table itself has gone from updating to active.

jsteinich replied:
It could complicate things, but at the cost of an extra description request, it could also just call `TableWaitOnStatusAsync`.
I unfortunately don't recall if the table status actually changed when adding a GSI to know if it's necessary.

Combining vs leaving separate seems like a wash, but maybe I'm missing something.

smiron replied:
> It could complicate things, but at the cost of an extra description request, it could also just call `TableWaitOnStatusAsync`. I unfortunately don't recall if the table status actually changed when adding a GSI to know if it's necessary.
> 
> Combining vs leaving separate seems like a wash, but maybe I'm missing something.

Doing the above would basically guarantee that the tables has gone  through an update. I'll check if adding a GSI triggers a table status change.

jsteinich reviewed (changes requested):
I didn't run a test, but compared to some very similar code that we have for handling table updates.

Really only thing that definitely stood out. Rest are more minor optimizations.

smiron commented:
> I didn't run a test, but compared to some very similar code that we have for handling table updates.
> 
> Really only thing that definitely stood out. Rest are more minor optimizations.

@jsteinich thanks for the review. I'll look into addressing the comments.
I'm not sure I understand what stood out and needs changing. Could you add more details please?

jsteinich commented:
> @jsteinich thanks for the review. I'll look into addressing the comments. I'm not sure I understand what stood out and needs changing. Could you add more details please?

Apologies, the provisioned throughput comment was the one definitive.

smiron commented:
> 

No worries. Thanks. I'll have that fixed.

smiron commented:
> > @jsteinich thanks for the review. I'll look into addressing the comments. I'm not sure I understand what stood out and needs changing. Could you add more details please?
> 
> Apologies, the provisioned throughput comment was the one definitive.

galvesribeiro reviewed (approved):
LGTM

smiron commented:
> > @jsteinich thanks for the review. I'll look into addressing the comments. I'm not sure I understand what stood out and needs changing. Could you add more details please?
> 
> Apologies, the provisioned throughput comment was the one definitive.

@jsteinich I've addressed your comments and some more corner cases. Please have another look.

jsteinich commented on src/AWS/Shared/Storage/DynamoDBStorage.cs:
```diff
@@ -65,9 +69,10 @@ internal class DynamoDBStorage
-            this.readCapacityUnits = readCapacityUnits;
-            this.writeCapacityUnits = writeCapacityUnits;
             this.useProvisionedThroughput = useProvisionedThroughput;
+            this.provisionedThroughput = this.useProvisionedThroughput
```
Nice

jsteinich commented on src/AWS/Shared/Storage/DynamoDBStorage.cs:
```diff
@@ -163,34 +165,191 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+            // This is because updating the TTL attribute requires (1) disabling the table TTL and (2) re-enabling it with the new TTL attribute.
+            // As per the below details page for this API: "It can take up to one hour for the change to fully process. Any additional UpdateTimeToLive calls for the same table during this one hour duration result in a ValidationException."
+            // https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateTimeToLive.html
+            if (describeTimeToLive.TimeToLiveStatus != TimeToLiveStatus.DISABLED)
```
Going from enabled to disabled seems feasible, but it definitely goes beyond this PR's scope.

smiron replied:
Indeed, definetly possible. The logic however gets quite complicated. I think the infrastructure provisioning code should be converted to leverage IaC (CDK or Terraform CDK would work well here). This would significantly reduce complexity.

jsteinich reviewed (approved):
Looks good. Thanks for all the effort.
